docs/design/chart_format.md

```diff
@@ -0,0 +1,664 @@
+
+
+The `templates/` directory contains one or more template files, as defined in the [Directory Layout](#directory-layout) section. Templates, along with all of their supporting files, are located therein.
+
+
+An optional `README.md` file may be specified, which contains long-form text information about using this chart. Tools may display this information, if present. The README.md file is in Markdown format, and should contain information about a Chart’s purpose, usage, and development.
+
+
+An optional `LICENSE` file may be present, which specifies license information for this chart and/or the images dependent on it.
+
+
+An optional `docs/` directory may be present, and may contain one or more documentation files. This directory contains documentation that is more specific, verbose, or thorough than what is present in the `README.md` file.
+
+
+An optional `hooks/` directory may be present. **This is reserved for future use.** A deployment engine may specify lifecycle hooks, such as `pre-install`, `pre-render`, `post-install` and `post-render`. Lifecycle hooks provide an opportunity for a chart author to run additional steps in the processing pipeline. Implementations of these hooks are stored here. _No other files may be stored in this directory._
+
+### Directory Layout
+A chart is laid out as follows. The top level directory (represented by the placeholder ROOT) must be the name of the chart. For example, if the chart is named `nginx`, the ROOT directory must be named `nginx/`.
+
+```
+ROOT/
+    Chart.yaml
+    README.md
+    LICENSE
+    docs/
+        some.md
+    templates/
+        some.yaml
+        some.jinja
+        some.jinja.schema
+```
+
+Templates are stored in a separate directory for the following reasons:
+
+
+* This future-proofs the format: we can add other directories at the top level and deprecate `templates/`.
+* It allows authors to add other files (such as documentation) in a straightforward way that will not confuse definitions-aware tools.
+* It allows for the possibility that a definition may be embedded inside of project code.
+
+
+Charts must not assume that they can perform file system operations to load another chart or supporting resources directly via the filesystem, nor should they store any operational files outside of the chart directory. It should be possible to archive the chart directory (e.g. `tar -cf ROOT.tar ROOT/`) and not lose any necessary information.
+
+### Generators and Templates
+The contents inside of the `templates/` directory may contain directives that can be processed by an external processor. Such a processor may read these files and generate new content.
+
+
+A generator reads through the files in `templates/` and looks for special file headers that indicate what generation tool needs to be run (if any) on a file. One such implementation of a generation tool is a template renderer. In any one pass through the `templates/` directory, more than one generation tool may be invoked.
```

Comment:
Why not use chart.yaml to declare a single generator that would be given the templates directory (and expected / forced to only read files from that directory).  That makes it harder to mix languages within a chart, but that doesn't seem like an obvious use case to me.  On the other hand it does allow a generator to simply "act on every file" or to for generators with imports (Python, Jinja, Jsonnet), pick an idiomatic name like main.py and use that as the entrypoint (or even let the chart.yaml specify the entrypoint).

It kinda feels like without a top-down execution plan, it might be hard for a human to tell how the templates/ dir would be interpreted.